packages/SwingSet/src/liveslots/virtualObjectManager.js

```diff
@@ -639,11 +642,47 @@ export function makeVirtualObjectManager(
   }
 
   function defineKind(tag, init, actualize, finish) {
-    return defineKindInternal(tag, init, actualize, finish, false);
+    const kindID = `${allocateExportID()}`;
+    return defineKindInternal(kindID, tag, init, actualize, finish, false);
+  }
+
+  let kindIDID;
```

Comment:
I think there's going to be a problem with this and upgrade, overlapping with #4730. If we run this code a second time in version-2 of a vat, where the value of `kindIDID` was already allocated by version-1 (and stored on disk), this implementation will allocate a new value, and overwrite the one on disk. And I think a version-2 which starts by defining a new durable kind, then unserializes something containing a Kind Handle, might not recognize those earlier ones on disk, because by the time `unserialize` happens, the meta kind is registered under a non-matching ID.

I'm thinking that during `startVat` we need to check the DB for the stored `kindIDID`, and used the stored value if present.

This won't cause us a problem today, but it'll cause problems when we implement upgrade, and I'd like to avoid as many roadblocks as we can anticipate.

Basically, we're going to have trouble with any counter we define only in RAM, but which is used for identifiers that can survive a vat upgrade, which includes all exported vrefs, and anything used in durable data. We need a general pattern of:
* long-term stable DB key name to hold the allocated value
* conditional initialization during `vatStart`
* we can get away with holding a copy in RAM during a single version's lifetime, but we need the latest value to be written to the DB before any upgrade
  * which might mean writing it to DB after every increment, or maybe we rely upon some sort of `bringOutYourDead` (but that feels fragile to me)

Reply:
On first read-thru I didn't understand this comment at all, until I looked and saw that the problem is actually `allocateExportID` itself (i.e., that it's a RAM-only counter) and not so much the way it's used here.  It's also used in a bunch of other places and all of those are going to have the same kind of problem with restart-after-upgrade.  I wrote the above code on the (mistaken) belief that `allocateExportID` kept a persistent counter (we have a lot of those; I had misremembered and thought it was one also).  `allocateExportID` definitely should keep its allocation counter in the database.  We already have quite a few DB-based counters that write on increment, so I don't think one more will have a substantial performance impact.

Reply:
Given that the vulnerability extends beyond durable kinds, I'm going to treat this as a separable issue and give it its own ticket rather than block landing this PR. #4855

Reply:
And it looks like issue #4730 that @warner cited is already that issue.  I'll close the new one as a dup.